doc/sphinx/practical-tools/utilities.rst

```diff
@@ -7,32 +7,307 @@
+project because it makes it easy to track the history of your changes.
+`git <https://git-scm.com/>`_ is the package most used by Coq projects.
+Typically, each package has its own git repository.
 
-::
+**Single directory projects.**  If your project has a single file (and
+depends only on packages in the Coq installation), put it in any directory
+you like and run Coq from that directory.  If your project has multiple files,
+then you need to compile them in an order that matches their dependencies.
+If B.v depends on A.v, then you should compile A.v before B.v.  You can do this
+with :n:`coqc A.v B.v`, but you may find it tedious to identify the dependencies.
+You may find it easier to treat your project as a multi-directory project.
 
-   % ocamlfind ocamlopt -thread -rectypes -linkall -linkpkg \
-                 -package coq.toplevel,coq.plugins.ltac \
-                 topbin/coqtop_bin.ml -o my_toplevel.native
+**Multi-directory projects.**  For these, we recommend using a `_CoqProject` file
+to define one or more packages for your project.  With that, you can automatically
+generate a makefile that matches the dependencies.
+A *package* is a group of files under a top directory and its subdirectories
+that's built and installed as a unit, just like the opam packages mentioned above.
+
+Large projects that have multiple packages (either their own code or dependencies
+on packages that you've not installed as user-contributed packages) need
+additional setup.  The `COQPATH` environment variable is good mechanism for such
+projects.  HYPERLINK to COQPATH or just below?
+
+Scripts in `.v` files must be compiled to `.vo` files using `coqc` before they
+can be :cmd:`Require`\d in other files.  Currently, each `.vo` file is created in
+the same directory as its `.v` file.  Furthermore, Coq has to know the logical
+path to associate with the directories containing your compiled files.
+
+.. _building_with_coqproject:
+
+Building a package with _CoqProject
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The _CoqProject file contains the information needed to generate a makefile
+for building your project.  Your _CoqProject file should be in
+the top directory of your package's source tree.  We recommend using the
+:term:`logical name` of the package as the name of the top directory.  Also,
+avoid putting your code into subdirectories such as "src" unless you want
+that to be part of the :term:`logical path` in Coq.
+
+For example, here's a minimal _CoqProject file for the `MyProject` package
+(the logical name of the package), which has 2 script files whose logical paths
+are `MyProject.File1` and `MyProject.SubDir.File2`.  `_CoqProject` should be
+in the top-level directory of your project.  We recommend naming that directory
+`MyProject`, the same as the logical name::
+
+  -R . MyPackage
+  File1.v
+  SubDir/File2.v
+
+TODO PR: Should we recommend -R (easier but more likely to give ambiguous paths) or
+-Q (safer, less likely to give ambiguous paths)?
```

Comment:
Yeah.  I'm learning and figuring out lots of things as I write this.

I think `-Q` is not ideal for referring to an external project because the associated directory is system-dependent.  (It will certainly be incorrect if publish your development in opam.)  You don't need it and shouldn't use it to refer to something that's installed, which shouldn't be re-compiled.  It would be good to have another flag to identify what packages the project depends on that wouldn't affect compilation, e.g. `-D Bignums` for `coq-bignums` or maybe `-D Bignums coq-bignums`.  For the future.

Reply:
The reality is that today, you do not need to set -Q and -R flags for other projects. Today, I can install mathcomp via opam and start using it in my own Coq project. I won't need to set -Q or -R or know anything about how mathcomp is installed. All I have to do is write `From mathcomp Require Import ...`, Coq will be able to find the file I specify no problem.

Internally, Coq knows that user-contrib is where it should start looking for things. Every directory in user-contrib gets `-Q` set for it automatically. That way the Coq command `From my_installed_dep Require someFile` can always be resolved.

Reply:
Don't worry about having to learn all these details when writing these docs. The details are awfully complicated and I am still learning a lot of new things every time I look into the details. The fact that we are even attempting to document this behaviour is a huge step in the right direction.

Reply:
> The reality is that today, you do not need to set -Q and -R flags for other projects.

Not needed for installed projects such as mathcomp, not needed if you use COQPATH for packages outside of the installation directory.  Maybe a useful shortcut if you don't use COQPATH.  Not that I love COQPATH...

> Don't worry about having to learn all these details when writing these docs.

It's hard write anything useful without considerable understanding of the material.  I write stuff, then figure out that it not right, then I redo it.  Sometime multiple iterations.  As long as it's right in the end (or at least clear enough to be correctible).

Reply:
I meant that: Of course you need to know what is going on to write about it, but don't worry about not understanding what is going on since there details are complex and most of us barely understand it anyway.

Reply:
Sure you could use COQPATH to also direct Coq to find the compiled vo files, however this is more of an "installation thing". If you are actively developing a multi-subproject project, you probably don't want to be "installing" parts of it for it to work.

Reply:
@SkySkimmer Would you mind correcting me then?

Reply:
-R is like -Q but allows require without an explicit prefix